File: src/core/phaseMachine.ts
/**
 * Pure state machine for the two-phase flow — Phaser-free so it unit-tests in plain Node.
 *
 * PhaseDirector feeds it three inputs and acts on the returned commands:
 *  - 'depleted'  (Zone A's buffer hit 0 and the board settled)  → A begins panning down.
 *  - 'barFilled' (Zone B's score bar cashed in)                 → B begins panning up.
 *  - 'panDone'   (the camera tween landed)                      → arrive at the target phase.
 *
 * Quirks encoded here (and covered by tests):
 *  - 'barFilled' in the A phase is IGNORED as a pan trigger: a cash-in can happen while
 *    already in A (e.g. milestone-drain balls fill the bar) — Zone A just refills in place.
 *  - 'barFilled' during A_TO_B can't turn the pan around mid-tween; it queues, and the
 *    machine bounces straight back (B → B_TO_A) the moment the downward pan lands.
 */
import type { GamePhase } from './contracts';

export interface PhaseState {
  phase: GamePhase;
  /** A cash-in arrived while panning down; bounce back up as soon as we land in B. */
  refillQueued: boolean;
}

export interface PhaseStep {
  state: PhaseState;
  /** Start the camera pan toward this phase (undefined = no new pan). */
  startPan?: 'B' | 'A';
  /** The phase changed — broadcast PHASE_CHANGED with the new state.phase. */
  changed: boolean;
}

export type PhaseInput = 'depleted' | 'barFilled' | 'panDone';

export function initialPhaseState(): PhaseState {
  return { phase: 'A', refillQueued: false };
}

export function stepPhase(state: PhaseState, input: PhaseInput): PhaseStep {
  const { phase, refillQueued } = state;

  switch (input) {
    case 'depleted':
      if (phase !== 'A') return { state, changed: false };
      return { state: { phase: 'A_TO_B', refillQueued }, startPan: 'B', changed: true };

    case 'barFilled':
      if (phase === 'B') {
        return { state: { phase: 'B_TO_A', refillQueued: false }, startPan: 'A', changed: true };
      }
      if (phase === 'A_TO_B') {
        return { state: { phase, refillQueued: true }, changed: false };
      }
      return { state, changed: false }; // in A (refill in place) or already panning up

    case 'panDone':
      if (phase === 'A_TO_B') {
        if (refillQueued) {
          // Landed in B with a refill already banked — turn straight around. B is never
          // observable here; the pan up starts the same tick.
          return { state: { phase: 'B_TO_A', refillQueued: false }, startPan: 'A', changed: true };
        }
        return { state: { phase: 'B', refillQueued }, changed: true };
      }
      if (phase === 'B_TO_A') {
        return { state: { phase: 'A', refillQueued }, changed: true };
      }
      return { state, changed: false };
  }
}

File: Assets/Game/Core/PhaseMachine.cs
namespace RichCoast.Core
{
    /// <summary>What the phase machine is fed (port of <c>phaseMachine.ts</c>).</summary>
    public enum PhaseInput
    {
        /// <summary>Zone A's buffer hit 0 and the board settled → A begins panning down.</summary>
        Depleted,
        /// <summary>Zone B's score bar cashed in → B begins panning up.</summary>
        BarFilled,
        /// <summary>The camera tween landed → arrive at the target phase.</summary>
        PanDone,
    }

    public readonly struct PhaseState
    {
        public readonly GamePhase Phase;
        /// <summary>A cash-in arrived while panning down; bounce back up as soon as we land in B.</summary>
        public readonly bool RefillQueued;

        public PhaseState(GamePhase phase, bool refillQueued)
        {
            Phase = phase;
            RefillQueued = refillQueued;
        }
    }

    public readonly struct PhaseStep
    {
        public readonly PhaseState State;
        /// <summary>Start the camera pan toward this phase (A or B); null = no new pan.</summary>
        public readonly GamePhase? StartPan;
        /// <summary>The phase changed — broadcast <c>PhaseChanged</c> with <see cref="State"/>.Phase.</summary>
        public readonly bool Changed;

        public PhaseStep(PhaseState state, GamePhase? startPan, bool changed)
        {
            State = state;
            StartPan = startPan;
            Changed = changed;
        }
    }

    /// <summary>
    /// Pure state machine for the two-phase flow. Quirks encoded here (and covered by tests):
    /// <c>BarFilled</c> in the A phase is IGNORED as a pan trigger (a cash-in can happen while already
    /// in A — Zone A just refills in place); <c>BarFilled</c> during A→B can't turn the pan around
    /// mid-tween — it queues, and the machine bounces straight back (B → B→A) the moment the
    /// downward pan lands.
    /// </summary>
    public static class PhaseMachine
    {
        public static PhaseState Initial => new PhaseState(GamePhase.A, false);

        public static PhaseStep Step(PhaseState state, PhaseInput input)
        {
            var phase = state.Phase;
            bool queued = state.RefillQueued;
            switch (input)
            {
                case PhaseInput.Depleted:
                    if (phase != GamePhase.A) return Same(state);
                    return new PhaseStep(new PhaseState(GamePhase.AToB, queued), GamePhase.B, true);

                case PhaseInput.BarFilled:
                    if (phase == GamePhase.B) return new PhaseStep(new PhaseState(GamePhase.BToA, false), GamePhase.A, true);
                    if (phase == GamePhase.AToB) return new PhaseStep(new PhaseState(phase, true), null, false);
                    return Same(state); // in A (refill in place) or already panning up

                case PhaseInput.PanDone:
                    if (phase == GamePhase.AToB)
                    {
                        if (queued) return new PhaseStep(new PhaseState(GamePhase.BToA, false), GamePhase.A, true);
                        return new PhaseStep(new PhaseState(GamePhase.B, queued), null, true);
                    }
                    if (phase == GamePhase.BToA) return new PhaseStep(new PhaseState(GamePhase.A, queued), null, true);
                    return Same(state);
            }
            return Same(state);
        }

        static PhaseStep Same(PhaseState state) => new PhaseStep(state, null, false);
    }
}
